database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala

```diff
@@ -74,14 +76,29 @@ trait WorkflowStoreEntryComponent {
     * Returns up to "limit" startable workflows, sorted by submission time.
     */
   val fetchStartableWorkflows = Compiled(
-    (limit: ConstColumn[Long]) => {
+    (limit: ConstColumn[Long], cromwellId: Rep[Option[String]], heartbeatThreshold: ConstColumn[Timestamp]) => {
       val query = for {
-        workflowStoreEntryRow <- workflowStoreEntries
-        if (workflowStoreEntryRow.workflowState === WorkflowStoreState.Aborting && workflowStoreEntryRow.restarted === true) ||
-           (workflowStoreEntryRow.workflowState === WorkflowStoreState.Running && workflowStoreEntryRow.restarted === true) ||
-           (workflowStoreEntryRow.workflowState === WorkflowStoreState.Submitted && workflowStoreEntryRow.restarted === false)
-      } yield workflowStoreEntryRow
-      query.sortBy(_.submissionTime.asc).take(limit)
+        row <- workflowStoreEntries
+        // This looks for:
+        //
+        // 1) Restarted workflows belonging to this Cromwell.
+        // 2) Submitted workflows not belonging to any Cromwell.
+        // 3) Workflows apparently orphaned by another Cromwell.
+        //
+        // Workflows are taken by submission time, oldest first. This is a "query for update", meaning rows are
+        // locked such that readers are blocked since we will do an update subsequent to this select in the same
+        // transaction that we know will impact those readers.
+        //
+        // The current code only writes heartbeats on initial pickup, so any other Cromwell's workflows will appear
+        // to be abandoned if they were picked up before `heartbeatThreshold`.
+        if (row.cromwellId === cromwellId && row.heartbeatTimestamp.isEmpty) || // 1
+          (row.cromwellId.isEmpty && row.heartbeatTimestamp.isEmpty) || // 2
+          (row.cromwellId =!= cromwellId && row.heartbeatTimestamp < heartbeatThreshold) // 3
+        // This logic leaves a hole where if another Cromwell comes up, nulls out the heartbeats of the workflows
+        // it had in flight, then goes down and stays down without picking up some of those workflows, no other
```

Comment:
The startup code nulls out the heartbeat so that this code will know the workflow is eligible for pickup. If the startup code wrote a current heartbeat then this code would wait `heartbeatTtl` before picking it up.